From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Sat, 19 Jan 2019 07:59:17 +0900
Subject: wait-online: do not fail if we receive invalid messages

Fixes #11486.

(cherry picked from commit 61a38e02650b8e7f097cadaa40aab0847605a383)
(cherry picked from commit 5d8935266c1ed546b7e869d2dd9fcb42303f9a7a)
---
 src/network/wait-online/manager.c | 31 ++++++++++++++++++-------------
 1 file changed, 18 insertions(+), 13 deletions(-)

diff --git a/src/network/wait-online/manager.c b/src/network/wait-online/manager.c
index 655fa0a..b7675cd 100644
--- a/src/network/wait-online/manager.c
+++ b/src/network/wait-online/manager.c
@@ -95,16 +95,25 @@ static int manager_process_link(sd_netlink *rtnl, sd_netlink_message *mm, void *
         assert(mm);
 
         r = sd_netlink_message_get_type(mm, &type);
-        if (r < 0)
-                goto fail;
+        if (r < 0) {
+                log_warning_errno(r, "rtnl: Could not get message type, ignoring: %m");
+                return 0;
+        }
 
         r = sd_rtnl_message_link_get_ifindex(mm, &ifindex);
-        if (r < 0)
-                goto fail;
+        if (r < 0) {
+                log_warning_errno(r, "rtnl: Could not get ifindex from link, ignoring: %m");
+                return 0;
+        } else if (ifindex <= 0) {
+                log_warning("rtnl: received link message with invalid ifindex %d, ignoring", ifindex);
+                return 0;
+        }
 
         r = sd_netlink_message_read_string(mm, IFLA_IFNAME, &ifname);
-        if (r < 0)
-                goto fail;
+        if (r < 0) {
+                log_warning_errno(r, "rtnl: Received link message without ifname, ignoring: %m");
+                return 0;
+        }
 
         l = hashmap_get(m->links, INT_TO_PTR(ifindex));
 
@@ -116,16 +125,16 @@ static int manager_process_link(sd_netlink *rtnl, sd_netlink_message *mm, void *
 
                         r = link_new(m, &l, ifindex, ifname);
                         if (r < 0)
-                                goto fail;
+                                return log_error_errno(r, "Failed to create link object: %m");
 
                         r = link_update_monitor(l);
                         if (r < 0)
-                                goto fail;
+                                return log_error_errno(r, "Failed to initialize link object: %m");
                 }
 
                 r = link_update_rtnl(l, mm);
                 if (r < 0)
-                        goto fail;
+                        return log_warning_errno(r, "Failed to process RTNL link message: %m");;
 
                 break;
 
@@ -139,10 +148,6 @@ static int manager_process_link(sd_netlink *rtnl, sd_netlink_message *mm, void *
         }
 
         return 0;
-
-fail:
-        log_warning_errno(r, "Failed to process RTNL link message: %m");
-        return 0;
 }
 
 static int on_rtnl_event(sd_netlink *rtnl, sd_netlink_message *mm, void *userdata) {
